### PR TITLE
AVRO-3955: [Rust] allow deserializer to decode enum string

### PR DESCRIPTION
**Pull request by FlorentinDUBOIS**
## What is the purpose of the change

This pull request allow to decode an avro enumeration from a previously encoded one.
Below, the example case to reproduce it.

```rust
    #[derive(PartialEq, Eq, Serialize, Deserialize, Debug)]
    pub struct StringEnum {
        pub source: String,
    }

    #[test]
    fn avro_xxxx_decode_enum() {
        let schema_content = r#"
{
  "name": "AccessLog",
  "namespace": "com.clevercloud.accesslogs.common.avro",
  "type": "record",
  "fields": [
    {
      "name": "source",
      "type": {
        "type": "enum",
        "name": "SourceType",
        "items": "string",
        "symbols": ["SOZU", "HAPROXY", "HAPROXY_TCP"]
      }
    }
  ]
}
"#;

        let schema = crate::Schema::parse_str(schema_content).unwrap();
        let data = StringEnum { source: "SOZU".to_string() };

        // encode into avro
        let value = crate::to_value(&data).unwrap();

        let mut buf = std::io::Cursor::new(crate::to_avro_datum(&schema, value).unwrap());

        // decode from avro
        let value = crate::from_avro_datum(&schema, &mut buf, None).unwrap();

        let decoded_data: StringEnum = crate::from_value(&value).unwrap();
        // :arrow_double_up: throw => Failed to deserialize Avro value into value: Expected a String|Bytes|Fixed|Uuid|Union, but got Enum(0, "SOZU")

        assert_eq!(decoded_data, data);
    }
```

## Verifying this change

This change is already covered by existing tests on enums and new tests `avro_3955_decode_enum` and `avro_3955_encode_enum_data_with_wrong_content`.

## Documentation

- Does this pull request introduce a new feature? no
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: lang/rust/avro/src/de.rs [hunk 1]
```diff
@@ -676,6 +676,49 @@ mod tests {
...
+        // encode into avro
+        let value = crate::to_value(&data).unwrap();
+
+        let mut buf = std::io::Cursor::new(crate::to_avro_datum(&schema, value).unwrap());
+
+        // decode from avro
+        let value = crate::from_avro_datum(&schema, &mut buf, None).unwrap();
+
+        let decoded_data: StringEnum = crate::from_value(&value).unwrap();
+        // :arrow_double_up: throw => Failed to deserialize Avro value into value: Expected a String|Bytes|Fixed|Uuid|Union, but got Enum(0, "SOZU")
```
Comment (martin-g) [inline lang/rust/avro/src/de.rs:717]: What is the purpose of this comment here ?

File: lang/rust/avro/src/de.rs [hunk 2]
```diff
@@ -676,6 +676,49 @@ mod tests {
...
+        "items": "string",
+        "symbols": ["SOZU", "HAPROXY", "HAPROXY_TCP"]
+      }
+    }
+  ]
+}
+"#;
+
+        let schema = crate::Schema::parse_str(schema_content).unwrap();
+        let data = StringEnum { source: "SOZU".to_string() };
```
Comment (martin-g) [inline lang/rust/avro/src/de.rs:706]: Let's add one more test case with a value that is not one of the enum symbols too!

File: lang/rust/avro/src/de.rs (hunk 1 above)
Comment (FlorentinDUBOIS) [reply lang/rust/avro/src/de.rs:717]: Just a forgotten comment from the issue example

File: lang/rust/avro/src/de.rs (hunk 2 above)
Comment (FlorentinDUBOIS) [reply lang/rust/avro/src/de.rs:706]: It is done :heavy_check_mark:

**Comment by martin-g** [comment]
Please create the Jira issue and update the PR title with it and the test method names!
I can update the commit messages when squash+merging the PR!

**Comment by martin-g** [comment]
You will need to `cargo fmt --all` too.

**Comment by FlorentinDUBOIS** [comment]
Will do :)

**Comment by FlorentinDUBOIS** [comment]
@martin-g the issue is created and it is [AVRO-3955](https://issues.apache.org/jira/browse/AVRO-3955). 
I have updated the pull request as well with the issue identifier.

File: lang/rust/avro/src/de.rs [hunk 3]
```diff
@@ -676,6 +676,85 @@ mod tests {
...
+            source: "SOZU".to_string(),
+        };
+
+        // encode into avro
+        let value = crate::to_value(&data).unwrap();
+
+        let mut buf = std::io::Cursor::new(crate::to_avro_datum(&schema, value).unwrap());
+
+        // decode from avro
+        let value = crate::from_avro_datum(&schema, &mut buf, None).unwrap();
```
Comment (martin-g) [inline lang/rust/avro/src/de.rs:716]:
```suggestion
        let value = crate::from_avro_datum(&schema, &mut buf, None)?;
```

File: lang/rust/avro/src/de.rs [hunk 4]
```diff
@@ -676,6 +676,85 @@ mod tests {
...
+        "name": "SourceType",
+        "items": "string",
+        "symbols": ["SOZU", "HAPROXY", "HAPROXY_TCP"]
+      }
+    }
+  ]
+}
+"#;
+
+        let schema = crate::Schema::parse_str(schema_content).unwrap();
```
Comment (martin-g) [inline lang/rust/avro/src/de.rs:744]:
```suggestion
        let schema = crate::Schema::parse_str(schema_content)?;
```

File: lang/rust/avro/src/de.rs [hunk 5]
```diff
@@ -676,6 +676,85 @@ mod tests {
...
+        let data = StringEnum {
+            source: "WRONG_ITEM".to_string(),
+        };
+
+        // encode into avro
+        let value = crate::to_value(data).unwrap();
```
Comment (martin-g) [inline lang/rust/avro/src/de.rs:750]:
```suggestion
        let value = crate::to_value(data)?;
```

**Comment by martin-g** [comment]
For some reason I cannot apply the suggestions from the Github UI ... I can do it later in the IDE!

**Comment by FlorentinDUBOIS** [comment]
I have updated the commit with your changes.
I think you could not edit the pull request as you have not the right to do it on my fork, but should be fine.

**Comment by FlorentinDUBOIS** [comment]
@martin-g can you click to trigger the ci?

**Comment by martin-g** [comment]
Thank you for the contribution, @FlorentinDUBOIS !

**Comment by FlorentinDUBOIS** [comment]
My pleasure :tophat:! Thank you very much! :heart:

**Pull request merged**